src/utils/bitfield.rs

```diff
@@ -0,0 +1,231 @@
+}
+
+macro_rules! impl_bitfields {
+    (for $($t:ty),+) => {
+        $(impl BitField for $t {
+            const TYPE_SIZE: usize = core::mem::size_of::<Self>() * 8;
+
+            fn get_bit(self, idx: usize) -> bool {
+                if idx >= Self::TYPE_SIZE {
+                    panic!("Index out of range for bit fields");
+                }
+
+                (self & (1 << idx)) != 0
+            }
+
+            fn get_bits<T: RangeBounds<usize>>(self, range: T) -> Self {
+                let range = to_regular_range(&range, Self::TYPE_SIZE);
+
+                if range.start >= Self::TYPE_SIZE || range.end > Self::TYPE_SIZE {
+                    panic!("Range out of range for bit fields");
+                }
+                if range.start >= range.end {
+                    panic!("End bound should be greater than lower one for bit fields.");
+                }
+
+                // Shift away high bits then lower ones
+                let val = self << (Self::TYPE_SIZE - range.end);
+                val >> ((Self::TYPE_SIZE - range.end) + range.start)
+            }
+
+            fn set_bit(self, idx: usize, value: bool) -> Self {
+                if idx >= Self::TYPE_SIZE {
+                    panic!("Index out of range for bit fields");
+                }
+
+                if value { self | (1 << idx) } else { self & (!(1 << idx)) }
+            }
+
+            fn set_bits<T: RangeBounds<usize>>(self, range: T, value: Self) -> Self {
+                let range = to_regular_range(&range, Self::TYPE_SIZE);
+
+                if range.start >= Self::TYPE_SIZE || range.end > Self::TYPE_SIZE {
+                    panic!("Range out of range for bit fields");
+                }
+                if range.start >= range.end {
+                    panic!("End bound should be greater than lower one for bit fields.");
+                }
+                if (range.end - range.start) < Self::TYPE_SIZE && (value >> (range.end - range.start)) > 0 {
+                    panic!("Value bigger than range.");
+                }
+
+                let mask: Self = !(!0 << (Self::TYPE_SIZE - range.end) >>
+                                    (Self::TYPE_SIZE - range.end) >>
+                                    range.start << range.start);
+                (self & mask) | (value << range.start)
+            }
+        })*
+    }
+}
+
+impl_bitfields!(for u8, u16, u32, u64);
+
+fn to_regular_range<T: RangeBounds<usize>>(range: &T, maximun: usize) -> Range<usize> {
+    let start = match range.start_bound() {
+        Bound::Excluded(&value) => value + 1,
+        Bound::Included(&value) => value,
+        Bound::Unbounded => 0,
+    };
+    let end = match range.end_bound() {
+        Bound::Excluded(&value) => value,
+        Bound::Included(&value) => value + 1,
+        Bound::Unbounded => maximun,
+    };
+
+    start..end
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // ############################################## get_bit
```

Comment:
We can rename this module `get_tests` and create another one named `set_tests` for example. This would prevent such comment and allow some editors to wrap the text conveniently.


Note that if we create a getter trait we might simply create it in a submodule and we could keep the standard `tests` name for test modules.

Reply:
Went for two tests modules in 16ae4c55091fec8209072769b7111929e14b1606